gcloud-java-pubsub/src/main/java/com/google/gcloud/pubsub/spi/PublisherApi.java

```diff
@@ -0,0 +1,498 @@
+    "https://www.googleapis.com/auth/pubsub"
+  };
+
+  public static final PathTemplate PROJECT_PATH_TEMPLATE =
+      PathTemplate.create("/projects/{project}");
+  public static final PathTemplate TOPIC_PATH_TEMPLATE =
+      PathTemplate.create("/projects/{project}/topics/{topic}");
+
+  // ========
+  // Members
+  // ========
+
+  private final ManagedChannel channel;
+  private final ServiceApiSettings settings;
+
+  // ===============
+  // Factory Methods
+  // ===============
+
+  /**
+   * Constructs an instance of PublisherApi with default settings.
+   */
+  public static PublisherApi create() throws IOException {
+    return create(new ServiceApiSettings());
+  }
+
+  /**
+   * Constructs an instance of PublisherApi, using the given settings. The channels are created based
+   * on the settings passed in, or defaults for any settings that are not set.
+   */
+  public static PublisherApi create(ServiceApiSettings settings) throws IOException {
+    return new PublisherApi(settings);
+  }
+
+  private PublisherApi(ServiceApiSettings settings) throws IOException {
+    ServiceApiSettings internalSettings = ApiUtils.settingsWithChannels(settings,
+        SERVICE_ADDRESS, DEFAULT_SERVICE_PORT, ALL_SCOPES);
+    this.settings = internalSettings;
+    this.channel = internalSettings.getChannel();
+  }
+
+  // ==============================
+  // Resource Name Helper Functions
+  // ==============================
+
+  public static final String createProject(String project) {
+    return PROJECT_PATH_TEMPLATE.instantiate(
+        "project", project);
+  }
+
+  public static final String createTopic(String project, String topic) {
+    return TOPIC_PATH_TEMPLATE.instantiate(
+        "project", project,"topic", topic);
+  }
+
+
+  // ========
+  // Getters
+  // ========
+
+  public Channel getChannel() {
+    return channel;
+  }
+
+
+  // =============
+  // Service Calls
+  // =============
+
+  // ----- createTopic -----
+
+  // AUTO-GENERATED DOCUMENTATION AND METHOD - see instructions at the top of the file for editing.
+  /**
+   * Creates the given topic with the given name.
+   *
+   * @param name The name of the topic. It must have the format
+   * `"projects/{project}/topics/{topic}"`. `{topic}` must start with a letter,
+   * and contain only letters (`[A-Za-z]`), numbers (`[0-9]`), dashes (`-`),
+   * underscores (`_`), periods (`.`), tildes (`~`), plus (`+`) or percent
+   * signs (`%`). It must be between 3 and 255 characters in length, and it
+   * must not start with `"goog"`.
+   */
+  public Topic createTopic(String name) {
+    Topic request =
+        Topic.newBuilder()
+        .setName(name)
+        .build();
+
+    return createTopic(request);
+  }
+
+  // AUTO-GENERATED DOCUMENTATION AND METHOD - see instructions at the top of the file for editing.
+  /**
+   * Creates the given topic with the given name.
+   *
+   * @param request The request object containing all of the parameters for the API call.
+   */
```

Comment:
I agree with you that the most important ones from a client perspective are `retryMaxAttempts` and `totalRetryPeriodMillis`. I think that most of the other retry settings tuning came from recommendation by the service owners themselves (e.g. they may have different values
for `maxRetryDelayMillis` and `initialRetryDelayMillis` to avoid hammering a service which
is facing issues).

Why can't we expose these with good defaults?